README.md
Sweater
===
Sweater is a program written in PHP designed to reverse engineer Club Penguin

### Requirements
<ul>
	<li> PHP 5.5+</li>
	<li> PDO w/ MySQL driver enabled</li>
	<li> Sockets</li>
	<li> PCNTL (optional)</li>
	<li> cURL (optional, used for updating crumbs)</li>
	<li> MySQL server</li>
	<li> Working Internet connection (optional, for updating crumbs)</li>
</ul>

### Instructions:
<ul>
	<li> Make sure you have an AS2 media server - once you have one, set it up</li>
	<li> Make sure you have the latest version of PHP installed</li>
		<li> If you need help, go to <a href="http://rile5.com/topic/23649-how-to-install-php-55-windows-linux/">this</a> topic for a tutorial on how to install the latest PHP version</li>
	<li> Make sure you have a MySQL server running</li>
	<li> Create a database and import Install.sql</li>
	<li> Create a user account using INSERT</li>
	<li> Configure Server.conf</li>
	<li> Execute Runner.php with a server ID from the configuration as an argument</li>
</ul>

###Usage:
<code>php Runner.php ServerID [--update-crumbs]</code>
	
####Example:
<code>php Runner.php 1 --update-crumbs</code><br />
<code>php Runner.php 100<br />


Information/CHANGELOG
  ______                                     
 / _____)                    _               
( (____  _ _ _ _____ _____ _| |_ _____  ____ 
 \____ \| | | | ___ (____ (_   _) ___ |/ ___)
 _____) ) | | | ____/ ___ | | |_| ____| |    
(______/ \___/|_____)_____|  \__)_____)_|    

Written by Arthur-N

Legend
	+ Successfully implemented (tested & fully functional)
	^ Requires completion (but is functional for the most part)
	% Requires completion (and is not functional)
	& Hasn't been implemented yet (Standard Club Penguin protocol)
	! Hasn't been implemented yet (Custom protocol)
	@ Won't be implemented.. at least no time soon
	$ Structural implementation (has been implemented)
	~ Structural implementation (needs to be implemented)
	# Structural implementation (requires completion)

+ XT & XML packet handling
+ Hashing for the login mechanism
+ Hashing for the game mechanism
+ Server list generation (for the server list at login)
	+ Caching the server list
+ Waddling
+ Sending messages (includes SafeChat messages)
+ Emotes
+ Actions (dancing, etc)
+ Buddies (includes adding, finding, removing & fetching)
+ Ignores (includes adding, removing & fetching)
+ Room and igloo joining
+ Igloos including upgrades, floors, music, (room) furniture, opening & furniture adding
+ Command handling
+ Adding & updating items
+ Puck moving for the soccer field & ice rink
+ Snowballs
+ Jokes 
+ Tour guide messages

^ Puffles 
	+ Adoption, having in igloos, puffle-cards, running away, statistics, walking
	^ Bathing, feeding
	& Playing with in-game as well as the playing with within the player's igloo (actions)

% Stamp protocol

^ Games 
	^ Single-player (Hydro Hopper, Catchin' Waves, Astro Barrier, Thin Ice, Puffle Roundup)
		^ Colors (works on some games)
		& Game-over packet handling for adding coins
	& Multi-player (Find Four, Card-Jitsu, Manacala, Sled Racing, Treasure)
	
& Digging
& Newspapers

! Marriage / BFF system (need client)
! Moods (need client)

$ Trait usage for the handlers as well as the cryptographic methods
$ File loader changes which include recursive directory loading as well as not needing a trailing slash for specified directories
$ File loader can skip / ignore files
$ Process-control implementation (limited only to *NIX platforms)
	$ No longer using ticks as they are, or may be, deprecated in the near the future
$ Logger now uses constants for the logging types
	$ Make use of logging types, perhaps create log files for the logger to write to
$ Writing retries to prevent data loss (it won't always be prevented, but it's supposed to lessen the amount of times it happens)
$ Encapsulation for Client data
$ Using encapsulation for client data
$ File loader is now gone, we are now using namespaces!

# Make full use of the extended classes of the PDO abstraction layer (GameHandler.php)
# Make full use of custom exceptions
# "Plugin" system


November 8, 2013 3:32 PM
	* Proper documentation of the change-log begins
	


Crumbs/Rooms.php
<?php
// Don't edit this //
$arrRooms = array(
        100 => array(
                'name' => 'town',
                'member' => 'false',
                'game' => 'false',
                'intid' => '1',
        ),
        110 => array(
                'name' => 'coffee',
                'member' => 'false',
                'game' => 'false',
                'intid' => '2',
        ),
        111 => array(
                'name' => 'book',
                'member' => 'false',
                'game' => 'false',
                'intid' => '3',
        ),
        120 => array(
                'name' => 'dance',
                'member' => 'false',
                'game' => 'false',
                'intid' => '4',
        ),
        121 => array(
                'name' => 'lounge',
                'member' => 'false',
                'game' => 'false',
                'intid' => '5',
        ),
        122 => array(
                'name' => 'eco',
                'member' => 'false',
                'game' => 'false',
                'intid' => '6',
        ),
        130 => array(
                'name' => 'shop',
                'member' => 'false',
                'game' => 'false',
                'intid' => '7',
        ),
        200 => array(
                'name' => 'village',
                'member' => 'false',
                'game' => 'false',
                'intid' => '8',
        ),
        211 => array(
                'name' => 'agentlobbysolo',
                'member' => 'false',
                'game' => 'false',
                'intid' => '9',
        ),
        212 => array(
                'name' => 'agentlobbymulti',
                'member' => 'false',
                'game' => 'false',
                'intid' => '10',
        ),
        213 => array(
                'name' => 'agentvr',
                'member' => 'false',
                'game' => 'false',
                'intid' => '105',
        ),
        220 => array(
                'name' => 'lodge',
                'member' => 'false',
                'game' => 'false',
                'intid' => '11',
        ),
        221 => array(
                'name' => 'attic',
                'member' => 'false',
                'game' => 'false',
                'intid' => '12',
        ),
        230 => array(
                'name' => 'mtn',
                'member' => 'false',
                'game' => 'false',
                'intid' => '13',
        ),
        300 => array(
                'name' => 'plaza',
                'member' => 'false',
                'game' => 'false',
                'intid' => '14',
        ),
        310 => array(
                'name' => 'pet',
                'member' => 'false',
                'game' => 'false',
                'intid' => '15',
        ),
        320 => array(
                'name' => 'dojo',
                'member' => 'false',
                'game' => 'false',
                'intid' => '16',
        ),
        321 => array(
                'name' => 'dojoext',
                'member' => 'false',
                'game' => 'false',
                'intid' => '17',
        ),
        322 => array(
                'name' => 'dojohide',
                'member' => 'false',
                'game' => 'false',
                'intid' => '18',
        ),
        323 => array(
                'name' => 'agentcom',
                'member' => 'false',
                'game' => 'false',
                'intid' => '19',
        ),
        330 => array(
                'name' => 'pizza',
                'member' => 'false',
                'game' => 'false',
                'intid' => '20',
        ),
        340 => array(
                'name' => 'stage',
                'member' => 'false',
                'game' => 'false',
                'intid' => '21',
        ),
        400 => array(
                'name' => 'beach',
                'member' => 'false',
                'game' => 'false',
                'intid' => '22',
        ),
        410 => array(
                'name' => 'light',
                'member' => 'false',
                'game' => 'false',
                'intid' => '23',
        ),
        411 => array(
                'name' => 'beacon',
                'member' => 'false',
                'game' => 'false',
                'intid' => '24',
        ),
        420 => array(
                'name' => 'ship',
                'member' => 'false',
                'game' => 'false',
                'intid' => '25',
        ),
        421 => array(
                'name' => 'shiphold',
                'member' => 'false',
                'game' => 'false',
                'intid' => '26',
        ),
        422 => array(
                'name' => 'shipquarters',
                'member' => 'false',
                'game' => 'false',
                'intid' => '27',
        ),
        423 => array(
                'name' => 'shipnest',
                'member' => 'false',
                'game' => 'false',
                'intid' => '28',
        ),
        800 => array(
                'name' => 'dock',
                'member' => 'false',
                'game' => 'false',
                'intid' => '29',
        ),
        801 => array(
                'name' => 'forts',
                'member' => 'false',
                'game' => 'false',
                'intid' => '30',
        ),
        802 => array(
                'name' => 'rink',
                'member' => 'false',
                'game' => 'false',
                'intid' => '31',
        ),
        803 => array(
                'name' => 'agent',
                'member' => 'false',
                'game' => 'false',
                'intid' => '32',
        ),
        804 => array(
                'name' => 'boiler',
                'member' => 'false',
                'game' => 'false',
                'intid' => '33',
        ),
        805 => array(
                'name' => 'berg',
                'member' => 'false',
                'game' => 'false',
                'intid' => '34',
        ),
        806 => array(
                'name' => 'cave',
                'member' => 'false',
                'game' => 'false',
                'intid' => '35',
        ),
        807 => array(
                'name' => 'shack',
                'member' => 'false',
                'game' => 'false',
                'intid' => '36',
        ),
        808 => array(
                'name' => 'mine',
                'member' => 'false',
                'game' => 'false',
                'intid' => '37',
        ),
        809 => array(
                'name' => 'forest',
                'member' => 'false',
                'game' => 'false',
                'intid' => '38',
        ),
        810 => array(
                'name' => 'cove',
                'member' => 'false',
                'game' => 'false',
                'intid' => '39',
        ),
        811 => array(
                'name' => 'boxdimension',
                'member' => 'false',
                'game' => 'false',
                'intid' => '40',
        ),
        812 => array(
                'name' => 'dojofire',
                'member' => 'true',
                'game' => 'false',
                'intid' => '41',
        ),
        813 => array(
                'name' => 'cavemine',
                'member' => 'false',
                'game' => 'false',
                'intid' => '42',
        ),
        814 => array(
                'name' => 'lake',
                'member' => 'false',
                'game' => 'false',
                'intid' => '43',
        ),
        815 => array(
                'name' => 'underwater',
                'member' => 'true',
                'game' => 'false',
                'intid' => '44',
        ),
        851 => array(
                'name' => 'party1',
                'member' => 'false',
                'game' => 'false',
                'intid' => '45',
        ),
        852 => array(
                'name' => 'party2',
                'member' => 'true',
                'game' => 'false',
                'intid' => '46',
        ),
        853 => array(
                'name' => 'party3',
                'member' => 'true',
                'game' => 'false',
                'intid' => '47',
        ),
        854 => array(
                'name' => 'party4',
                'member' => 'true',
                'game' => 'false',
                'intid' => '48',
        ),
        855 => array(
                'name' => 'party5',
                'member' => 'true',
                'game' => 'false',
                'intid' => '49',
        ),
        856 => array(
                'name' => 'party6',
                'member' => 'true',
                'game' => 'false',
                'intid' => '50',
        ),
        857 => array(
                'name' => 'party7',
                'member' => 'true',
                'game' => 'false',
                'intid' => '51',
        ),
        858 => array(
                'name' => 'party8',
                'member' => 'true',
                'game' => 'false',
                'intid' => '52',
        ),
        859 => array(
                'name' => 'party9',
                'member' => 'true',
                'game' => 'false',
                'intid' => '53',
        ),
        860 => array(
                'name' => 'party10',
                'member' => 'true',
                'game' => 'false',
                'intid' => '54',
        ),
        861 => array(
                'name' => 'party11',
                'member' => 'true',
                'game' => 'false',
                'intid' => '55',
        ),
        862 => array(
                'name' => 'party12',
                'member' => 'true',
                'game' => 'false',
                'intid' => '56',
        ),
        863 => array(
                'name' => 'party13',
                'member' => 'true',
                'game' => 'false',
                'intid' => '57',
        ),
        864 => array(
                'name' => 'party14',
                'member' => 'true',
                'game' => 'false',
                'intid' => '58',
        ),
        865 => array(
                'name' => 'party15',
                'member' => 'true',
                'game' => 'false',
                'intid' => '59',
        ),
        866 => array(
                'name' => 'party16',
                'member' => 'true',
                'game' => 'false',
                'intid' => '60',
        ),
        867 => array(
                'name' => 'party17',
                'member' => 'true',
                'game' => 'false',
                'intid' => '61',
        ),
        868 => array(
                'name' => 'party18',
                'member' => 'true',
                'game' => 'false',
                'intid' => '62',
        ),
        899 => array(
                'name' => 'party',
                'member' => 'true',
                'game' => 'false',
                'intid' => '63',
        ),
        900 => array(
                'name' => 'astro',
                'member' => 'false',
                'game' => 'true',
                'intid' => '64',
        ),
        901 => array(
                'name' => 'beans',
                'member' => 'false',
                'game' => 'true',
                'intid' => '65',
        ),
        902 => array(
                'name' => 'roundup',
                'member' => 'false',
                'game' => 'true',
                'intid' => '66',
        ),
        903 => array(
                'name' => 'hydro',
                'member' => 'false',
                'game' => 'true',
                'intid' => '67',
        ),
        904 => array(
                'name' => 'fish',
                'member' => 'false',
                'game' => 'true',
                'intid' => '68',
        ),
        905 => array(
                'name' => 'cart',
                'member' => 'false',
                'game' => 'true',
                'intid' => '69',
        ),
        906 => array(
                'name' => 'jetpack',
                'member' => 'false',
                'game' => 'true',
                'intid' => '70',
        ),
        907 => array(
                'name' => 'mission1',
                'member' => 'false',
                'game' => 'true',
                'intid' => '71',
        ),
        908 => array(
                'name' => 'mission2',
                'member' => 'false',
                'game' => 'true',
                'intid' => '72',
        ),
        909 => array(
                'name' => 'thinice',
                'member' => 'false',
                'game' => 'true',
                'intid' => '73',
        ),
        910 => array(
                'name' => 'pizzatron',
                'member' => 'false',
                'game' => 'true',
                'intid' => '74',
        ),
        911 => array(
                'name' => 'mission3',
                'member' => 'false',
                'game' => 'true',
                'intid' => '75',
        ),
        912 => array(
                'name' => 'waves',
                'member' => 'false',
                'game' => 'true',
                'intid' => '76',
        ),
        913 => array(
                'name' => 'mission4',
                'member' => 'false',
                'game' => 'true',
                'intid' => '77',
        ),
        914 => array(
                'name' => 'mission5',
                'member' => 'false',
                'game' => 'true',
                'intid' => '78',
        ),
        915 => array(
                'name' => 'mission6',
                'member' => 'false',
                'game' => 'true',
                'intid' => '79',
        ),
        916 => array(
                'name' => 'aqua',
                'member' => 'false',
                'game' => 'true',
                'intid' => '80',
        ),
        917 => array(
                'name' => 'book1',
                'member' => 'false',
                'game' => 'true',
                'intid' => '81',
        ),
        918 => array(
                'name' => 'book2',
                'member' => 'false',
                'game' => 'true',
                'intid' => '82',
        ),
        919 => array(
                'name' => 'book3',
                'member' => 'false',
                'game' => 'true',
                'intid' => '83',
        ),
        920 => array(
                'name' => 'mission7',
                'member' => 'false',
                'game' => 'true',
                'intid' => '84',
        ),
        921 => array(
                'name' => 'mission8',
                'member' => 'false',
                'game' => 'true',
                'intid' => '85',
        ),
        922 => array(
                'name' => 'mission9',
                'member' => 'false',
                'game' => 'true',
                'intid' => '86',
        ),
        923 => array(
                'name' => 'mission10',
                'member' => 'false',
                'game' => 'true',
                'intid' => '87',
        ),
        926 => array(
                'name' => 'mixmaster',
                'member' => 'false',
                'game' => 'true',
                'intid' => '88',
        ),
        927 => array(
                'name' => 'mission11',
                'member' => 'false',
                'game' => 'true',
                'intid' => '89',
        ),
        941 => array(
                'name' => 'soaker',
                'member' => 'false',
                'game' => 'true',
                'intid' => '90',
        ),
        942 => array(
                'name' => 'balloon',
                'member' => 'false',
                'game' => 'true',
                'intid' => '91',
        ),
        943 => array(
                'name' => 'bell',
                'member' => 'false',
                'game' => 'true',
                'intid' => '92',
        ),
        944 => array(
                'name' => 'feed',
                'member' => 'false',
                'game' => 'true',
                'intid' => '93',
        ),
        945 => array(
                'name' => 'memory',
                'member' => 'false',
                'game' => 'true',
                'intid' => '94',
        ),
        946 => array(
                'name' => 'paddle',
                'member' => 'false',
                'game' => 'true',
                'intid' => '95',
        ),
        947 => array(
                'name' => 'shuffle',
                'member' => 'false',
                'game' => 'true',
                'intid' => '96',
        ),
        948 => array(
                'name' => 'spin',
                'member' => 'false',
                'game' => 'true',
                'intid' => '97',
        ),
        949 => array(
                'name' => 'rescue',
                'member' => 'false',
                'game' => 'true',
                'intid' => '98',
        ),
        951 => array(
                'name' => 'sensei',
                'member' => 'false',
                'game' => 'true',
                'intid' => '99',
        ),
        952 => array(
                'name' => 'dancing',
                'member' => 'false',
                'game' => 'true',
                'intid' => '100',
        ),
        953 => array(
                'name' => 'firesensei',
                'member' => 'false',
                'game' => 'true',
                'intid' => '101',
        ),
        997 => array(
                'name' => 'fire',
                'member' => 'false',
                'game' => 'true',
                'intid' => '102',
        ),
        998 => array(
                'name' => 'card',
                'member' => 'false',
                'game' => 'true',
                'intid' => '103',
        ),
        999 => array(
                'name' => 'sled',
                'member' => 'false',
                'game' => 'true',
                'intid' => '104',
        ),
);
?>
